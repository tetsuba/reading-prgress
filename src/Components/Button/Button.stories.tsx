import type { Meta, StoryObj } from '@storybook/react'
import Button from './Button'

const meta = {
    title: 'COMPONENTS/Button',
    component: Button,
    parameters: {
        layout: 'centered'
    },
    tags: ['autodocs'],
    args: {
        template: 'primary',
        children: 'text'
    },
    argTypes: {
        className: {
            control: 'text',
            description: 'Add tailwind classes for positioning'
        },
        children: {
            description: '',
            options: ['None', 'Text'],
            control: { type: 'select' },
            mapping: {
                None: '',
                Text: 'Text'
            }
        }
    }
} satisfies Meta<typeof Button>

export default meta
type Story = StoryObj<typeof meta>

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Primary: Story = {
    args: {
        template: 'primary'
    }
}

export const Secondary: Story = {
    args: {
        template: 'secondary',
        children: 'text'
    }
}

export const Tertiary: Story = {
    args: {
        template: 'tertiary',
        children: 'text'
    }
}

export const Warning: Story = {
    args: {
        template: 'warning',
        children: 'text'
    }
}

export const Disabled: Story = {
    args: {
        template: 'disabled',
        children: 'text'
    }
}

export const TextLink: Story = {
    args: {
        template: 'text-link',
        children: 'text'
    }
}

export const IconCloseButton: Story = {
    args: {
        icon: 'close',
        template: 'icon-close',
        children: ''
    }
}

export const IconHistoryButton: Story = {
    args: {
        icon: 'history',
        template: 'icon-history',
        children: ''
    }
}

export const IconBackButton: Story = {
    args: {
        icon: 'back',
        template: 'icon-back',
        children: ''
    }
}

export const IconUserButton: Story = {
    args: {
        icon: 'user',
        template: 'icon-user',
        children: ''
    }
}

export const IconDeleteButton: Story = {
    args: {
        icon: 'delete',
        template: 'icon-delete',
        children: ''
    }
}

export const IconCheckBadgeButton: Story = {
    args: {
        icon: 'check-badge',
        template: 'icon-check-badge',
        children: ''
    }
}

export const IconLeftAndTextButton: Story = {
    args: {
        icon: 'back',
        template: 'primary',
        children: <span className="ml-2">Text</span>
    }
}

export const IconRightAndTextButton: Story = {
    args: {
        icon: 'back',
        template: 'primary',
        children: <span className="mr-2">Text</span>,
        right: true
    }
}
