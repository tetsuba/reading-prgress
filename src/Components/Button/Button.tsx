import Svg, { SvgTypes } from '../Svg/Svg'

type ChildrenTypes = string | JSX.Element
type ButtonTemplateTypes = 'icon' | 'primary' | 'none' | 'secondary' | 'warning' | 'tertiary'

type PropTypes = {
    children?: ChildrenTypes | ChildrenTypes[]
    clickHandler?: () => void
    id?: string
    type?: 'button' | 'submit' | 'reset'
    svg?: SvgTypes
    className?: string
    dataTestid?: string
    template: ButtonTemplateTypes
    title?: string
}

const defaultStyles = 'rounded-lg font-medium text-base text-center px-5 py-2.5'

export const buttonClasses: { [k: string]: string } = {
    none: '',
    primary: `${defaultStyles} text-white bg-blue-700 hover:bg-blue-800`,
    secondary: `${defaultStyles} text-gray bg-gray-200 hover:bg-gray-300`,
    tertiary: `${defaultStyles} text-white bg-green-500 hover:bg-green-600`,
    warning: `${defaultStyles} text-white bg-red-700 hover:bg-red-800`,
    icon: 'bg-transparent inline-flex items-center rounded-lg text-sm text-gray-400'
}

export default function Button(props: PropTypes) {
    const buttonClassNames = buttonClasses[props.template]
        ? buttonClasses[props.template]
        : ''

    const classNames = props.className ? props.className : ''

    return (
        <button
            title={props.title}
            data-testid={props.dataTestid}
            onClick={props.clickHandler}
            className={`${buttonClassNames} ${classNames}`}
            id={props.id}
            type={props.type}
        >
            {props.svg ? <Svg type={props.svg} /> : props.children}
        </button>
    )
}
