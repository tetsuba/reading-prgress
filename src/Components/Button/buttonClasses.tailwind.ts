const BUTTON_CLASSES = {
    FILLED: {
        DEFAULT:
            'flex items-center rounded-lg font-medium text-base text-center px-5 py-2.5',
        get PRIMARY() {
            return `${this.DEFAULT} text-white bg-blue-700 hover:bg-blue-800`
        },
        get SECONDARY() {
            return `${this.DEFAULT} text-gray bg-gray-200 hover:bg-gray-300`
        },
        get TERTIARY() {
            return `${this.DEFAULT} text-white bg-green-500 hover:bg-green-600`
        },
        get WARNING() {
            return `${this.DEFAULT} text-white bg-red-700 hover:bg-red-800`
        },
        get DISABLED() {
            return `${this.DEFAULT} text-gray-400 bg-gray-200`
        }
    },
    TEXT_LINK: 'text-blue-700 hover:underline dark:text-blue-500',
    ICON: {
        DEFAULT:
            'bg-transparent inline-flex items-center rounded-lg text-sm text-gray-400',
        get CLOSE() {
            return `${this.DEFAULT} p-1.5`
        },
        get USER() {
            return `${this.DEFAULT} rounded-full bg-gray-800`
        },
        get HISTORY() {
            return `${this.DEFAULT} p-2`
        },
        get BACK() {
            return `${this.DEFAULT} p-2`
        }
    },
    HOVER: {
        GREY: 'hover:bg-gray-200 hover:text-gray-900 dark:hover:bg-gray-800 dark:hover:text-white'
        // GREY: 'hover:bg-gray-100 hover:text-gray-900 focus:outline-none hover:border-white '
    },
    get primary(): string {
        return `${this.FILLED.PRIMARY}`
    },
    get secondary() {
        return `${this.FILLED.SECONDARY}`
    },
    get tertiary() {
        return `${this.FILLED.TERTIARY}`
    },
    get warning() {
        return `${this.FILLED.WARNING}`
    },
    get disabled() {
        return `${this.FILLED.DISABLED}`
    },
    get svgClose() {
        return `${this.ICON.CLOSE} ${this.HOVER.GREY}`
    },
    get ['icon-close']() {
        return `${this.ICON.CLOSE} ${this.HOVER.GREY}`
    },
    get ['icon-history']() {
        return `${this.ICON.HISTORY} ${this.HOVER.GREY}`
    },
    get ['icon-back']() {
        return `${this.ICON.BACK} ${this.HOVER.GREY}`
    },
    get ['icon-user']() {
        return `${this.ICON.USER}`
    },
    get ['icon-check-badge']() {
        return `${this.ICON.DEFAULT} hover:text-green-500`
    },
    get ['icon-delete']() {
        return `${this.ICON.DEFAULT} ${this.HOVER.GREY}`
    },
    get icon() {
        return `${this.ICON.DEFAULT}`
    },
    get ['text-link']() {
        return this.TEXT_LINK
    },
    ['nav-menu-desktop']:
        'block w-full bg-white px-4 py-2 text-left text-sm text-gray-700 hover:border-white hover:bg-gray-200',
    readingWord:
        'mr-5 inline-block cursor-pointer border-b-2 transition-all duration-300 ',
    ['menu-button-mobile']:
        'w-full bg-blue-500 py-4 text-center text-2xl text-white',
    none: ''
}

export type TailwindTemplateTypes =
    | 'primary'
    | 'secondary'
    | 'tertiary'
    | 'warning'
    | 'disabled'
    | 'text-link'
    | 'icon-close'
    | 'icon-history'
    | 'icon-back'
    | 'icon-user'
    | 'icon-check-badge'
    | 'icon-delete'
    | 'icon'
    | 'none'
    | 'nav-menu-desktop'
    | 'readingWord'
    | 'menu-button-mobile'
export default function getTailWindClasses(key: TailwindTemplateTypes): string {
    return BUTTON_CLASSES[key]
}
