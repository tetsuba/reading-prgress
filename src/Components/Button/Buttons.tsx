import Svg, { SvgIconTypes } from '../Svg/Svg'
import Button from './Button'

type PropTypes = {
    onClick?: () => void
}

export function BackToBooksButton(props: PropTypes) {
    return (
        <Button
            className="ml-4 flex items-center place-self-start"
            data-testid="back-button"
            template="secondary"
            {...props}
        >
            <Svg icon="back" />
            <span className="ml-2 hidden md:inline">Back to books</span>
        </Button>
    )
}
