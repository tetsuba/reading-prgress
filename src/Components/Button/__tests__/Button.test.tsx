import { fireEvent, render, screen } from '@testing-library/react'
import Button, { ButtonTemplateTypes } from '../Button'

describe('Button', () => {
    const buttonTemplates: ButtonTemplateTypes[] = [
        'none',
        'primary',
        'secondary',
        'tertiary',
        'warning'
    ]
    test.each(buttonTemplates)(
        'should render %s button',
        (template: ButtonTemplateTypes) => {
            const { asFragment } = render(
                <Button type="button" template={template}>
                    {template}
                </Button>
            )
            expect(asFragment()).toMatchSnapshot()
        }
    )
    test('should render an icon button', () => {
        const { asFragment } = render(
            <Button
                title="title"
                data-testid="test"
                className="m-2"
                type="button"
                template="icon"
                svg="warning"
            />
        )
        expect(asFragment()).toMatchSnapshot()
    })
})
