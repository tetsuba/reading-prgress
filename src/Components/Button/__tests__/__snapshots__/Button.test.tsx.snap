// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > should render an icon button 1`] = `
<DocumentFragment>
  <button
    class="bg-transparent inline-flex items-center rounded-lg text-sm text-gray-400 p-1.5 hover:bg-gray-200 hover:text-gray-900 dark:hover:bg-gray-800 dark:hover:text-white "
  >
    <svg
      aria-hidden="true"
      class="h-5 w-5"
      fill="currentColor"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clip-rule="evenodd"
        d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
        fill-rule="evenodd"
      />
    </svg>
  </button>
</DocumentFragment>
`;

exports[`Button > should render primary button 1`] = `
<DocumentFragment>
  <button
    class="flex items-center rounded-lg font-medium text-base text-center px-5 py-2.5 text-white bg-blue-700 hover:bg-blue-800 "
  >
    primary
  </button>
</DocumentFragment>
`;

exports[`Button > should render secondary button 1`] = `
<DocumentFragment>
  <button
    class="flex items-center rounded-lg font-medium text-base text-center px-5 py-2.5 text-gray bg-gray-200 hover:bg-gray-300 "
  >
    secondary
  </button>
</DocumentFragment>
`;

exports[`Button > should render svgClose button 1`] = `
<DocumentFragment>
  <button
    class="bg-transparent inline-flex items-center rounded-lg text-sm text-gray-400 p-1.5 hover:bg-gray-200 hover:text-gray-900 dark:hover:bg-gray-800 dark:hover:text-white "
  >
    svgClose
  </button>
</DocumentFragment>
`;

exports[`Button > should render tertiary button 1`] = `
<DocumentFragment>
  <button
    class="flex items-center rounded-lg font-medium text-base text-center px-5 py-2.5 text-white bg-green-500 hover:bg-green-600 "
  >
    tertiary
  </button>
</DocumentFragment>
`;

exports[`Button > should render textLink button 1`] = `
<DocumentFragment>
  <button
    class="undefined "
  >
    textLink
  </button>
</DocumentFragment>
`;

exports[`Button > should render warning button 1`] = `
<DocumentFragment>
  <button
    class="flex items-center rounded-lg font-medium text-base text-center px-5 py-2.5 text-white bg-red-700 hover:bg-red-800 "
  >
    warning
  </button>
</DocumentFragment>
`;
